bftengine/src/bftengine/DbCheckpointManager.cpp

```diff
@@ -115,14 +116,62 @@ void DbCheckpointManager::loadCheckpointDataFromPersistence() {
 }
 
 void DbCheckpointManager::init() {
-  // check if there is chkpt data in persistence
-  loadCheckpointDataFromPersistence();
-  // if there is a checkpoint created in database but entry corresponding to that is not found in persistence
-  // then probably, its a partially created checkpoint and we want to remove it
-  auto listOfCheckpointsCreated = dbClient_->getListOfCreatedCheckpoints();
-  for (auto& cp : listOfCheckpointsCreated) {
-    if (dbCheckptMetadata_.dbCheckPoints_.find(cp) == dbCheckptMetadata_.dbCheckPoints_.end()) {
-      removeCheckpoint(cp);
+  // In case of scale with restart we have to update dbcheckpoint metadata from file system
+  // if any dbcheckpoints are present in the file system.
+  if (lastCheckpointSeqNum_ == 0 && EpochManager::instance().getSelfEpochNumber() > 0) {
+    // update metadata for dbcheckpoints from file system
+    const auto& checkpointDir = dbClient_->getCheckpointPath();
+    _fs::path path(checkpointDir);
+    std::vector<_fs::directory_entry> allDBCheckpoints;
+
+    try {
+      if (_fs::exists(path)) {
+        for (const auto& entry : _fs::directory_iterator(checkpointDir)) {
+          const auto filenameStr = entry.path().filename().string();
+          if (_fs::is_directory(entry)) {
+            allDBCheckpoints.push_back(entry);
+            const CheckpointId& checkPointId = std::stoi(filenameStr);
+            const auto blockId = checkPointId;
+
+            auto p = path / filenameStr;
+            _fs::file_time_type ftime = _fs::last_write_time(p);
+            std::time_t cftime = decltype(ftime)::clock::to_time_t(ftime);
+            const std::chrono::seconds creationTime = std::chrono::seconds(cftime);
+            {
+              std::scoped_lock lock(lockLastDbCheckpointDesc_);
+              lastCreatedCheckpointMetadata_.emplace(DbCheckpointMetadata::DbCheckPointDescriptor{
+                  checkPointId, creationTime, blockId, lastCheckpointSeqNum_});
+              dbCheckptMetadata_.dbCheckPoints_.insert({checkPointId, lastCreatedCheckpointMetadata_.value()});
+            }
+            numOfDbCheckpointsCreated_++;
+          }
+        }
+
+        // extract the last created BlockId from all created dbCheckpoints.
+        auto lastCreatedDbCheckpoint =
+            *std::max_element(allDBCheckpoints.begin(),
+                              allDBCheckpoints.end(),
+                              [](const _fs::directory_entry& f1, const _fs::directory_entry& f2) {
+                                return _fs::last_write_time(f1.path()) < _fs::last_write_time(f2.path());
+                              });
+
+        lastDbCheckpointBlockId_.Get().Set(std::stoi(lastCreatedDbCheckpoint.path().filename()));
+        metrics_.UpdateAggregator();
+      }
+      updateMetrics();
+    } catch (std::exception& e) {
+      LOG_WARN(getLogger(), "Failed to find any checkpoints: " << e.what());
```

Comment:
Warning message could be: failed to update checkpoints metadata from checkpoint directory